File: android_cursor_grok/data/crypto/src/main/kotlin/com/transcripto/crypto/VigenereCipher.kt
package com.transcripto.crypto

import com.transcripto.domain.CipherMethod

class VigenereCipher : CipherMethod {
    override fun encrypt(input: String, params: Map<String, Any>): String {
        val key = prepareKey(params["key"] as? String ?: "", params["salt"] as? String ?: "")
        if (key.isEmpty()) throw IllegalArgumentException("Clave inválida")
        return process(input, key, true)
    }

    override fun decrypt(input: String, params: Map<String, Any>): String {
        val key = prepareKey(params["key"] as? String ?: "", params["salt"] as? String ?: "")
        if (key.isEmpty()) throw IllegalArgumentException("Clave inválida")
        return process(input, key, false)
    }

    private fun prepareKey(key: String, salt: String): String {
        val effectiveKey = key + salt
        return effectiveKey.filter { it.isLetter() }
    }

    private fun process(text: String, key: String, encrypt: Boolean): String {
        val result = StringBuilder()
        var keyIndex = 0
        for (char in text) {
            if (char.isLetter()) {
                val shift = key[keyIndex % key.length].uppercaseChar().code - 'A'.code
                val shifted = shiftChar(char, if (encrypt) shift else -shift)
                result.append(shifted)
                keyIndex++
            } else {
                result.append(char)
            }
        }
        return result.toString()
    }

    private fun shiftChar(c: Char, shift: Int): Char {
        val base = if (c.isUpperCase()) 'A' else 'a'
        return ((c.code - base.code + shift + 26) % 26 + base.code).toChar()
    }
}

File: android_flutter/transcripto/packages/data_crypto/lib/src/vigenere_cipher.dart
import 'dart:convert';

import 'package:domain/domain.dart';
import 'package:core_common/core_common.dart';

class VigenereCipher implements CipherMethod {
  @override
  String encrypt(String input, CipherParams params) {
    final key = _effectiveKey(params);
    if (key.isEmpty) {
      throw ArgumentError('Clave inválida');
    }
    final k = key;
    final sb = StringBuffer();
    int ki = 0;
    for (final ch in input.split('')) {
      final code = ch.codeUnitAt(0);
      if (_isAlpha(code)) {
        final isUpper = code >= 65 && code <= 90;
        final base = isUpper ? 65 : 97;
        final shift = _charShift(k[ki % k.length]);
        final pos = code - base;
        final npos = (pos + shift) % 26;
        sb.writeCharCode(base + npos);
        ki++;
      } else {
        sb.write(ch);
      }
    }
    return sb.toString();
  }

  @override
  String decrypt(String input, CipherParams params) {
    final key = _effectiveKey(params);
    if (key.isEmpty) {
      throw ArgumentError('Clave inválida');
    }
    final k = key;
    final sb = StringBuffer();
    int ki = 0;
    for (final ch in input.split('')) {
      final code = ch.codeUnitAt(0);
      if (_isAlpha(code)) {
        final isUpper = code >= 65 && code <= 90;
        final base = isUpper ? 65 : 97;
        final shift = _charShift(k[ki % k.length]);
        final pos = code - base;
        final npos = (pos - shift) % 26;
        sb.writeCharCode(base + (npos < 0 ? npos + 26 : npos));
        ki++;
      } else {
        sb.write(ch);
      }
    }
    return sb.toString();
  }

  String _normalizeLetters(String s) {
    final noDia = removeDiacritics(s);
    final lettersOnly = noDia.replaceAll(RegExp('[^A-Za-z]'), '');
    return lettersOnly;
  }

  String _effectiveKey(CipherParams p) {
    var key = _normalizeLetters(p.key ?? '');
    if (p.useSalt && p.salt != null) {
      final saltStr = _normalizeLetters(utf8.decode(p.salt!));
      key = key + saltStr;
    }
    return key;
  }

  bool _isAlpha(int code) {
    return (code >= 65 && code <= 90) || (code >= 97 && code <= 122);
  }

  int _charShift(String ch) {
    if (ch.isEmpty) return 0;
    final c = ch.codeUnitAt(0);
    if (c >= 65 && c <= 90) return c - 65;
    if (c >= 97 && c <= 122) return c - 97;
    return 0;
  }
}
